### DISCUSSION TITLE
Meaningful indentation?

### DISCUSSION THREAD

**Discussion by tslocke** [eclipse-langium/langium#608] [Q&A] [answered]
Hi, I'm working on a DSL that uses indentation to delimit block structure, like Python or Haskell. Is that something Langium can cope with?

I see you are using Chevrotain under the hood - my current implementation is also in Chevrotain, and makes use of the pattern matching callbacks in the lexer.

I've been putting off writing a VS Code editing mode, and if I understand right, Langium could make that a lot easier. My understanding was that VC Code syntax highlighting is driven not by a full parser but by a bunch of regexes, like in the old TextMate modes. Is that something Langium can derive from the grammar? Seems difficult!

**Comment by msujew** [comment]
Hi @tslocke, you're getting into uncharted waters here, as we haven't tried that out yet, but as long as it's possible in Chevrotain, it should be possible in Langium as well.

By default, Langium will create relatively simple `TokenType` objects from each keyword and terminal that's found in the Langium grammar definition of your language. All of them go through the [TokenBuilder](https://github.com/langium/langium/blob/d8da5a5b56643b559706263d1c6a7b40b5c1a448/packages/langium/src/parser/token-builder.ts#L40-L54) service, where you can fine-tune each `TokenType` object. You would probably need to create dummy `INDENT` and `DEDENT` tokens, reference those in your parser rules and add a pattern matching callback to them in your own `TokenBuilder` implementation.

> My understanding was that VC Code syntax highlighting is driven not by a full parser but by a bunch of regexes, like in the old TextMate modes.

vscode syntax highlighting is twofold: For one we can simply let vscode do syntactical highlighting, which is provided based on TextMate grammars. Additionally, the language server protocol also offers semantic highlighting, which enables far more fine-grained control over your highlighting.

> Is that something Langium can derive from the grammar? Seems difficult!

We actually already generate TextMate grammars from Langium grammars, see the [textmate-generator.ts](https://github.com/langium/langium/blob/main/packages/langium-cli/src/generator/textmate-generator.ts). It suffices for a lot of use cases, but isn't recommended for production use of more complicated languages, as these probably require a bit fine tuning. It does these kinds of highlighting all by itself:

* Keywords
* Comments
* Strings

All the TextMate grammars for our [example languages](https://github.com/langium/langium/blob/main/examples/domainmodel/syntaxes/domain-model.tmLanguage.json) are generated using the provided generator.

Semantic highlighting can be implemented using the [AbstractSemanticTokenProvider](https://github.com/langium/langium/blob/main/packages/langium/src/lsp/semantic-token-provider.ts) and allows to perform highlighting based on AST info.

**Reply by msujew** [reply]
@tslocke By the way the [Chevrotain documentation](https://chevrotain.io/docs/guide/custom_token_patterns.html) contains an example on how one would go on about [python style indentation lexing](https://github.com/chevrotain/chevrotain/blob/master/examples/lexer/python_indentation/python_indentation.js) using token callbacks.

**Comment by tslocke** [comment]
Thanks for the advice. Yes I already have it working in chevrotain based on that example. I think I'm firmly in the realm of "more complicated languages" in terms of generating a useful textmate grammar.

**Reply by aabounegm** [reply]
Hi @tslocke, can you please share a link to your working implementation, if it's open-source? I am also currently trying to implement a language with indentation-delimited blocks and struggling a bit to implement it correctly.

**Reply by msujew** [reply]
@aabounegm See https://github.com/eclipse-langium/langium/discussions/1573#discussioncomment-10009009 for an implementation of this.

**Reply by aabounegm** [reply]
@msujew Great! Thank you very much for the reference, seems like I simply had an incomplete RegEx and an issue with one `if` condition 😁.
Anyway, seeing as this issue is relatively common, I would like to contribute my solution so others don't have to copy the same Token Builder and Lexer to their projects. Would you prefer a PR to the Langium repo (simply adding the new implementation to be imported by users who need them) or a separate npm module with these 2 classes?

**Reply by msujew** [reply]
Sure, a PR would be appreciated, thanks! See also https://github.com/eclipse-langium/langium/issues/1016#issuecomment-1547325802 for a bit of inspiration on the API of the feature. You can directly link the issue in your PR, since that would effectively close the issue.

**Reply by aabounegm** [reply]
@msujew I submitted a PR now at #1578. Hope it helps

**Reply by aabounegm** [reply, marked as answer]
The [IndentationAwareTokenBuilder](https://eclipse-langium.github.io/langium/classes/langium.IndentationAwareTokenBuilder.html) has been released in v3.2.0, and there is now [a recipe](https://langium.org/docs/recipes/lexing/indentation-sensitive-languages/) explaining how to use it.
